### DISCUSSION TITLE
v.0.8 Beta

### DISCUSSION THREAD

**Discussion by ramokz** [ramokz/phantom-camera#398] [Announcements]
> [!important]
> This beta is intended as a pre-release build to allow the community to give feedback on the upcoming version before it goes live for everyone.
> The main area of feedback interest is about the Phantom Camera Noise feature (see below).

> [!Note]
> The version number in `plugin.cfg` is purposely set to `v0.7.9`.
> This is to allow for upgrading to `v0.8` from the Beta release using the addon's [updater](https://phantom-camera.dev/editor-updater).

# :rotating_light: Breaking Changes
- `append_look_at_targets_array` used to take a `Array[NodePath]` as a parameter
  - Parameter now only accepts type `Array[Node3D]`

# :warning: Deprecated Functions  & Properties
- `erase_look_at_targets_member` is now deprecated
  - Use `erase_look_at_targets` instead.

# :sparkles:  New Features
## Phantom Camera Noise
A long requested feature, camera noise, also known as camera shake, is available for `PCam2Ds` and `PCam3Ds`, able to affect both the rotation and position of the `PCam`.

<img src="https://github.com/user-attachments/assets/b4151820-9b16-4595-9193-fdff732adf28" width="80" />

### Phantom Camera Noise Resource
Named `PhantomCameraNoiseResource`, is a new `resource` type with a 2D and 3D variant. It defines all the variables for the noise such as intensity, rotational / position limits and randomness seed. The noise is calculated using a perlin noise pattern.

Both 2D and 3D `Noise Resources` can apply changes to `position` and `rotation`; however, it's recommended to avoid rotational noise in 3D scenes to avoid geometry clipping issues.

---

<img src="https://github.com/user-attachments/assets/e77bf9d4-6486-4531-8df4-cc77b051a3a2" width="80" />
<img src="https://github.com/user-attachments/assets/9550edc5-5afc-4bb9-8636-3ae1c792ba70" width="80" />

### Phantom Camera Noise Emitter
Named `PhantomCameraNoiseEmitter`, is a new `node` type with a 2D and 3D variant for their respective scenes. Its primary use case is to trigger noise at a given moment, by calling its `.emit()` method, for either a specified duration or a continous loop. It can be previewed in the editor by toggling the `Preview` property in the inspector.

Each emitter uses a `Noise Resource` to define its noise pattern and only affects the `PCams` in the matching `Noise Emitter Layers`. That is to say, in order for the emitter to work both the emitter and the individual `PCam` node needs to have _at least_ one matching `Noise Layer`

### Quick Guide
There are 2 ways to apply noise to a `Camera2D/3D` with this system:
1. Supply a `Noise Resource` to the `Noise` property in a given `PCam` directly. Doing so will make the noise run persistently without pause on that specific `PCam`. **Note:** This will only run after a tween has been completed. 
3. Add a `Noise Emitter` (`PhantomCameraNoiseEmitter` node) to the scene, assign a `Noise Resource` and trigger it at any time by calling its `emit()` function — again, it can be previewed in the viewfinder by toggling the `Preview` button. The Emitter can be placed anywhere in the scene tree. **Note:** This will impact the `Camera2D/3D` both when following _and_ during a tween to a `PCam` if it has a matching `Noise Layer`.

#### Using External Noise Values
If the user wants to supply their own noise values, either from third-party libraries or using their scripts, then that can be achieved by calling `emit_noise()` from a `PCam2D/3D` by supplying a `Transform2D` or `Transform3D` respectively. 

#### Examples Scenes
- 2D - `2d_noise_example_scene.tscn`
- 3D - `3d_noise_example_scene.tscn`

# :arrow_up:  Improvements
- TileMapLayer Support for Limit Node
  - Thanks @Jack-023 for the PR (#378)
- Simplified Follow (2D / 3D) and Look At (3D) logic in the tick thread
- Added `dead_zone_reached` signal, is emitted whenever the target touches the edge of the dead zone
- Improved the behaviour of 2D Framed Follow. Should now no longer change vertical camera movement when only the horizontal dead zones are reached and vice versa
  - Thanks @lafitteque for the PR (#373)

# :bug: Fixes
- Resolved a crash that would occur when freeing a parent of `follow_target`, `follow_targets` (for group follow) or `look_at_target` and `look_at_targets` (group look at) (#381)
  -  Thanks @TranquilMarmot for the report (#374)
- Resolved error spam when changing 2D scene
  - Thanks @Baba-Yagan for the PR (#376)
- Resolved error spam when changing 3D scene
- Resolved an issue with getting an error spam when having an empty or a valid target not in the first array index when using `Follow Group` or `Look At Group`

<hr /><em>This discussion was created from the release <a href='https://github.com/ramokz/phantom-camera/releases/tag/camera-noise-beta'>v.0.8 Beta</a>.</em>

**Comment by MigasEu** [comment]
Hello!
I was checking the beta to solve the jitter issue I had on 0.7, which it fixes, but I had to make a small change for it to work for me, which was not have the third person camera go the the root, but keep it on the same parent.
That's because I have a reference to the phantom camera to change the fov later.

```gdscript
_follow_spring_arm.margin = margin
_follow_spring_arm.add_excluded_object(follow_target)
#get_tree().root.add_child.call_deferred(_follow_spring_arm) (OLD)
get_parent().add_child.call_deferred(_follow_spring_arm)
reparent.call_deferred(_follow_spring_arm)
``` 

Is there a specific reason for this to change to the root, since the spring arm is still set as top level, anyway?

Thanks!

**Reply by ramokz** [reply]
The change was mainly just a cosmetic one so that the `spring_arm` would exist in the top level of the scene tree. But you're right that it has no practical effect due to the `top_level` property being set to `true` further above.

Haven't noticed any changes with this particular update myself, but am happy to shift it back to the `get_parent()` method if you're experiencing undesired side effects from the `get_tree()` change.

**Reply by MigasEu** [reply]
That was the only issue I had. An option to choose between the 2 options would also work.
It just makes it a lot more difficult to manage it at an individually scene instance, like I was before.

**Reply by ramokz** [reply]
Done! https://github.com/ramokz/phantom-camera/commit/560f3376096ebf8a6e39cf0fb9068ed3abbac554

**Reply by ramokz** [reply]
Thanks for the feedback, appreciate it

**Reply by MigasEu** [reply]
Thank you for changing it and for the add-on in general! 🙏
